plugins/node/instrumentation-fs/src/instrumentation.ts

```diff
@@ -0,0 +1,321 @@
+      }
+
+      const lastIdx = args.length - 1;
+      const cb = args[lastIdx];
+      if (typeof cb === 'function') {
+        const span = instrumentation.tracer.startSpan(
+          `fs ${fnName}`
+        ) as api.Span;
+
+        // return to the context active during the call in the callback
+        args[lastIdx] = api.context.bind(
+          api.context.active(),
+          function (this: unknown, err?: Error) {
+            if (err) {
+              span.recordException(err);
+              span.setStatus({
+                message: err.message,
+                code: api.SpanStatusCode.ERROR,
+              });
+            }
+            if (typeof endHook === 'function') {
+              try {
+                endHook(fnName, { args: arguments, span });
+              } catch (e) {
+                instrumentation._diag.error('caught endHook error', e);
+              }
+            }
+            span.end();
+            return cb.apply(this, arguments);
+          }
+        );
+
+        try {
+          return api.context.with(
+            api.trace.setSpan(api.context.active(), span),
+            original,
+            this,
+            ...args
+          );
+        } catch (err) {
+          span.recordException(err);
+          span.setStatus({
+            message: err.message,
+            code: api.SpanStatusCode.ERROR,
+          });
+          if (typeof endHook === 'function') {
+            try {
+              endHook(fnName, { args: args, span });
+            } catch (e) {
+              instrumentation._diag.error('caught endHook error', e);
+            }
+          }
+          span.end();
+          throw err;
+        }
+      } else {
+        // TODO: what to do if we are pretty sure it's going to throw
+        return original.apply(this, args);
+      }
+    };
+  }
+
+  protected _patchSyncFunction<T extends (...args: any[]) => ReturnType<T>>(
+    fnName: string,
+    original: T
+  ): T {
+    const instrumentation = this;
+    return <any>function (this: any, ...args: any[]) {
+      if (isTracingSuppressed(api.context.active())) {
+        // Performance optimization. Avoid creating additional contexts and spans
+        // if we already know that the tracing is being suppressed.
+        return original.apply(this, args);
+      }
+      const { createHook, endHook } =
+        instrumentation.getConfig() as FsInstrumentationConfig;
+      if (typeof createHook === 'function') {
+        if (
+          createHook(fnName, {
+            args: args,
+          }) === false
+        ) {
+          // return original.apply(this, args);
+          return api.context.with(
+            suppressTracing(api.context.active()),
+            original,
+            this,
+            ...args
+          );
+        }
+      }
+
+      const span = instrumentation.tracer.startSpan(`fs ${fnName}`) as api.Span;
+      try {
+        // QUESTION: Should we immediately suppress all internal nested calls?
```

Comment:
I think we should. We should at least be consistent with the case above when `createHook` return false

Reply:
Hmm... interesting. I don't see a strong reason to seek consistency with the `createHook` behavior since the scenarios are different.

1. The question in the comment concerns whether an user can expect to see only top-level calls which were initiated in the top level.
2. `createHook` behavior concerns the use-case when user has disabled the main call thus already signalling disinterest to the calls nested IMHO.

But you are right pointing out that `createHook` doesn't cover all the usecases: for example when I want to catch ALL(even nested) the "open" calls but nothing else.

I'd vote for
1. 70% removing suppressing nested calls and giving that control(and responsibility) to the user OR
2. 30% leaving it as is making it convenient to ignore noise.

I would want to avoid, however making the solution more complex than it absolutely has to be now and in the future. `fs` is very frequently called API and I would want to stay as slim as possible in such a hot path.